src/wasm/wasm-stack.cpp

```diff
@@ -2707,11 +2761,116 @@ InsertOrderedMap<Type, Index> BinaryInstWriter::countScratchLocals() {
+    // a more expensive analysis later, which we only run if we see something
+    // suspicious here.
+    Index numDangerousBrIfs = 0;
+
+    void visitBreak(Break* curr) {
+      if (curr->type.hasRef()) {
+        numDangerousBrIfs++;
+      }
+    }
+
+    void visitDrop(Drop* curr) {
+      if (curr->value->is<Break>() && curr->value->type.hasRef()) {
+        // The value is exactly a br_if of a ref, that we just visited before
+        // us. Undo the ++ from there as it can be ignored.
+        assert(numDangerousBrIfs > 0);
+        numDangerousBrIfs--;
+      }
+    }
+  } finder(*this);
   finder.walk(func->body);
 
+  if (!finder.numDangerousBrIfs || !parent.getModule()->features.hasGC()) {
+    // Nothing more to do: either no such br_ifs, or GC is not enabled.
+    //
+    // The explicit check for GC is here because if only reference types are
+    // enabled then we still may seem to need a fixup here, e.g. if a ref.func
+    // is br_if'd to a block of type funcref. But that only appears that way
+    // because in Binaryen IR we allow non-nullable types even without GC (and
+    // if GC is not enabled then we always emit nullable types in the binary).
+    // That is, even if we see a type difference without GC, it will vanish in
+    // the binary format; there is never a need to add any ref.casts without GC
+    // being enabled.
+    return std::move(finder.scratches);
+  }
+
+  // There are dangerous-looking br_ifs, so we must do the harder work to
+  // actually investigate them in detail, including tracking block types. By
+  // being fully precise here, we'll only emit casts when absolutely necessary,
+  // which avoids repeated roundtrips adding more and more code.
+  struct RefinementScanner : public ExpressionStackWalker<RefinementScanner> {
+    BinaryInstWriter& writer;
+    ScratchLocalFinder& finder;
+
+    RefinementScanner(BinaryInstWriter& writer, ScratchLocalFinder& finder)
+      : writer(writer), finder(finder) {}
+
+    void visitBreak(Break* curr) {
+      // See if this is one of the dangerous br_ifs we must handle.
+      if (!curr->type.hasRef()) {
+        // Not even a reference.
+        return;
+      }
+      auto* parent = getParent();
+      if (parent) {
+        if (parent->is<Drop>()) {
+          // It is dropped anyhow.
+          return;
+        }
+        if (auto* cast = parent->dynCast<RefCast>()) {
+          if (Type::isSubType(cast->type, curr->type)) {
+            // It is cast to the same type or a better one. In particular this
+            // handles the case of repeated roundtripping: After the first
+            // roundtrip we emit a cast that we'll identify here, and not emit
+            // an additional one.
+            return;
+          }
+        }
+      }
+      auto* breakTarget = findBreakTarget(curr->name);
+      auto unrefinedType = breakTarget->type;
+      if (unrefinedType == curr->type) {
+        // It has the proper type anyhow.
+        return;
+      }
```

Comment:
After #6579 there will be no need to know the source type when emitting a cast, so this check will be the only remaining use for `unrefinedType`. This is also the only check that cannot be done as-is in `ScratchLocalFinder`. WDYT about making `ScratchLocalFinder` an `ExpressionStackWalker` to avoid needing this separate walker at all?

Reply:
The overhead is annoying, I'm afraid, and I'd really like to not regress the by-far common case of not needing the extra work.

Some data:

```
$ perf stat -r 5 bin/wasm-opt UE4Game-HTML5-Shipping.wasm -all -o a.wasm
warning: no passes specified, not doing any work
warning: no passes specified, not doing any work
warning: no passes specified, not doing any work
warning: no passes specified, not doing any work
warning: no passes specified, not doing any work

 Performance counter stats for 'bin/wasm-opt UE4Game-HTML5-Shipping.wasm -all -o a.wasm' (5 runs):

         10,947.27 msec task-clock                       #    1.767 CPUs utilized               ( +-  0.38% )
               855      context-switches                 #   78.102 /sec                        ( +-  9.24% )
                40      cpu-migrations                   #    3.654 /sec                        ( +- 11.87% )
           239,470      page-faults                      #   21.875 K/sec                       ( +-  0.07% )
    34,826,556,461      cycles                           #    3.181 GHz                         ( +-  0.27% )
    48,135,304,544      instructions                     #    1.38  insn per cycle              ( +-  0.01% )
    10,101,559,953      branches                         #  922.747 M/sec                       ( +-  0.02% )
       150,681,797      branch-misses                    #    1.49% of all branches             ( +-  1.17% )

            6.1962 +- 0.0404 seconds time elapsed  ( +-  0.65% )
```

And, with this patch:
```diff
$ git diff
diff --git a/src/wasm/wasm-stack.cpp b/src/wasm/wasm-stack.cpp
index 53a25d52b..1a83f56cb 100644
--- a/src/wasm/wasm-stack.cpp
+++ b/src/wasm/wasm-stack.cpp
@@ -2681,7 +2681,7 @@ void BinaryInstWriter::noteLocalType(Type type, Index count) {
 }
 
 InsertOrderedMap<Type, Index> BinaryInstWriter::countScratchLocals() {
-  struct ScratchLocalFinder : PostWalker<ScratchLocalFinder> {
+  struct ScratchLocalFinder : ExpressionStackWalker<ScratchLocalFinder> {
     BinaryInstWriter& parent;
     InsertOrderedMap<Type, Index> scratches;
```
```
$ perf stat -r 5 bin/wasm-opt UE4Game-HTML5-Shipping.wasm -all -o b.wasm

 Performance counter stats for 'bin/wasm-opt UE4Game-HTML5-Shipping.wasm -all -o b.wasm' (5 runs):

         11,311.23 msec task-clock                       #    1.727 CPUs utilized               ( +-  0.26% )
               720      context-switches                 #   63.654 /sec                        ( +-  5.01% )
                29      cpu-migrations                   #    2.564 /sec                        ( +-  8.30% )
           238,817      page-faults                      #   21.113 K/sec                       ( +-  0.04% )
    36,257,517,592      cycles                           #    3.205 GHz                         ( +-  0.14% )
    50,906,632,590      instructions                     #    1.40  insn per cycle              ( +-  0.66% )
    10,525,197,435      branches                         #  930.509 M/sec                       ( +-  0.01% )
       155,907,135      branch-misses                    #    1.48% of all branches             ( +-  0.70% )

            6.5514 +- 0.0335 seconds time elapsed  ( +-  0.51% )
```

That's about 5% more instructions and walltime. Note that that measures load+save, so the impact on save is around double that.

Reply:
Oof, that's way too much slowdown :(